pytorch_pfn_extras/onnx/_lax.py

```diff
@@ -0,0 +1,623 @@
+"""
+This file provides APIs to define control-flow operators (e.g., onnx::Loop and onnx::If)
+when using `ppe.onnx.export`.
+`torch.jit` records only `first loop` during tracing, and ppe.onnx inserts control-flow
+operators after exporting ONNX.
+
+APIs are almost same as [jax.lax](https://jax.readthedocs.io/en/latest/jax.lax.html).
+"""
+
+import torch
+import threading
+import onnx
+import onnx.helper
+from typing import Generator, Callable, Any, List, Tuple, Union, Dict
+from contextlib import contextmanager
+
+from pytorch_pfn_extras.onnx._as_output import as_output
+
+
+_lax_state = threading.local()
+
+_torch_dtype_to_onnx_dtype_dict = {
+    torch.bool: onnx.TensorProto.BOOL,
+    torch.uint8: onnx.TensorProto.UINT8,
+    torch.int8: onnx.TensorProto.INT8,
+    torch.int16: onnx.TensorProto.INT16,
+    torch.int32: onnx.TensorProto.INT32,
+    torch.int64: onnx.TensorProto.INT64,
+    torch.float16: onnx.TensorProto.FLOAT16,
+    torch.float32: onnx.TensorProto.FLOAT,
+    torch.float64: onnx.TensorProto.DOUBLE,
+    torch.complex64: onnx.TensorProto.COMPLEX64,
+    torch.complex128: onnx.TensorProto.COMPLEX128,
+}
+
+
+@contextmanager
+def init_lax_state() -> Generator[None, None, None]:
+    _lax_state.n_call = 0
+    _lax_state.input_for_postproc = {}
+    _lax_state.ignore_trace = False
+    try:
+        yield
+    finally:
+        _lax_state.ignore_trace = None
+        _lax_state.n_call = None
+        _lax_state.input_for_postproc = None
+
+
+class _DummyOpForControlFlow(torch.autograd.Function):
+    @staticmethod
+    def forward(  # type: ignore
+        ctx: Any,
+        init_val: torch.Tensor,
+        actual_val: torch.Tensor,
+    ) -> torch.Tensor:
+        return actual_val
+
+    @staticmethod
+    def symbolic(g, init_val, actual_val):  # type: ignore
+        return init_val
+
+
+State = Union[torch.Tensor, Tuple[torch.Tensor, ...]]
+
+
+def _as_tuple(val: State) -> Tuple[torch.Tensor, ...]:
+    if isinstance(val, torch.Tensor):
+        return tuple([val])
+    else:
+        assert isinstance(val, tuple)
+        return val
+
+
+def _apply(val: State, f: Callable[[int, torch.Tensor], torch.Tensor]) -> State:
+    if isinstance(val, torch.Tensor):
+        return f(0, val)
+    else:
+        assert isinstance(val, tuple)
+        return tuple([f(i, v) for i, v in enumerate(val)])
+
+
+def _trace() -> bool:
+    if not torch.jit.is_tracing():  # type: ignore[no-untyped-call]
+        return False
+    if hasattr(_lax_state, "ignore_trace") and _lax_state.ignore_trace:
```

Comment:
Does this `hasattr` checking exist for direct call from `torch.onnx.export` ?
If so, how about warn about control flow won't  generated correctly